internal/graph/track_check_resolver.go

```diff
@@ -0,0 +1,211 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+const (
+	trackerInterval = time.Duration(60) * time.Second
+)
+
+// TrackerCheckResolverOpt defines an option pattern that can be used to change the behavior of TrackerCheckResolver.
+type TrackerCheckResolverOpt func(checkResolver *TrackerCheckResolver)
+type resolutionTree struct {
+	tm   time.Time
+	hits *atomic.Uint64
+}
+
+type TrackerCheckResolver struct {
+	delegate CheckResolver
+	ticker   *time.Ticker
+	logger   logger.Logger
+	ctx      context.Context
+	nodes    sync.Map
+}
+
+var _ CheckResolver = (*TrackerCheckResolver)(nil)
+
+func WithTrackerLogger(logger logger.Logger) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.logger = logger
+	}
+}
+
+func WithTrackerContext(ctx context.Context) TrackerCheckResolverOpt {
+	return func(t *TrackerCheckResolver) {
+		t.ctx = ctx
```

Comment:
So this is to listen to cancelations of the server itself and do graceful shutdown? I'm not seeing a final flush tho 🤔

Reply:
Yes, if the server terminates so will this go-routine and all associated stored entries. I will add a final flush of all entries to logs.